packages/loader/src/gltf/parser/GLTFParserContext.ts

```diff
@@ -1,67 +1,150 @@
+   * Get the resource of the specified type.
+   * @remarks Except for Entity, all other parsers are asynchronous
+   * @param type - The type of resource
+   * @param index - The index of resource, default all
    */
-  getBuffers(): Promise<ArrayBuffer[]> {
-    return Promise.resolve(this._buffers);
+  get<T>(type: GLTFParserType, index?: number): T {
+    const cache = this._resourceCache;
+    const isOnlyOne = [GLTFParserType.JSON, GLTFParserType.Validator].indexOf(type) !== -1;
+    let cacheKey: string;
+    let needOnlyOne = true;
+
+    if (isOnlyOne) {
+      cacheKey = `${type}`;
+    } else if (index >= 0) {
+      cacheKey = `${type}:${index}`;
+    } else {
+      cacheKey = `${type}`;
+      needOnlyOne = false;
+    }
+
+    let resource: T = cache.get(cacheKey);
+
+    if (!resource) {
+      if (needOnlyOne) {
+        resource = GLTFParserContext._parsers[type].parse(this, index);
+
+        // store glTFResource
+        if (!isOnlyOne) {
+          if (type === GLTFParserType.Entity) {
+            if (!this.glTFResource["entities"]) {
+              this.glTFResource["entities"] = [];
+            }
+
+            this.glTFResource["entities"][index] = <Entity>resource;
+          } else {
+            if (!this.glTFResource[type]) {
+              this.glTFResource[type] = [];
+            }
+            (<Promise<T>>resource)
+              .then((item: T) => {
+                if (!this.glTFResource[type]) {
+                  this.glTFResource[type] = [];
+                }
+
+                this.glTFResource[type][index] = item;
+
+                if (type === GLTFParserType.Mesh) {
+                  for (let i = 0, length = (<ModelMesh[]>item).length; i < length; i++) {
+                    const mesh = item[i] as ModelMesh;
+                    this.resourceManager.onSubAssetSuccess<ModelMesh>(`${this.url}?q=${type}[${index}][${i}]`, mesh);
+                  }
+                } else {
+                  this.resourceManager.onSubAssetSuccess<T>(`${this.url}?q=${type}[${index}]`, item);
+                  if (type === GLTFParserType.Scene && (this.glTF.scene ?? 0) === index) {
+                    this.resourceManager.onSubAssetSuccess<Entity>(`${this.url}?q=defaultSceneRoot`, item as Entity);
+                  }
+                }
+              })
+              .catch((e) => {
+                this.resourceManager.onSubAssetFail(`${this.url}?q=${type}[${index}]`, e);
+              });
+          }
+        }
+      } else {
+        const items = this.glTF[type];
+        if (items) {
+          if (type === GLTFParserType.Entity) {
+            resource = items.map((_, index) => this.get<T>(type, index));
+          } else {
+            resource = <T>Promise.all(items.map((_, index) => this.get<T>(type, index)));
+          }
+        } else {
+          resource = <T>Promise.resolve(null);
+        }
+      }
+
+      cache.set(cacheKey, resource);
+    }
+
+    return resource;
```

Comment:
Please improve the logical readability and type hints of the function！

Reply:
done